Add Debugging ROS 2 Router tutorial (#90)
eProsima/vulcanexus: docs/rst/tutorials/cloud/debug_ros2_router/debug_ros2_router.rst

.. include:: ../../../exports/alias.include

.. _tutorials_debug_ros2_router:

Debugging ROS 2 Router
======================

.. contents::
    :depth: 2
    :local:
    :backlinks: none

Background
----------

*eProsima ROS 2 Router*, a.k.a `DDS Router <https://github.com/eProsima/DDS-Router>`_, is an end-user software application that enables the connection of distributed ROS 2 networks (see *ROS 2 Router* documentation :ref:`here <vulcanexus_router>`).
That is, ROS 2 nodes such as publishers and subscriptions, or clients and services, deployed in one geographic location and using a dedicated local network will be able to communicate with other ROS 2 nodes deployed in different geographic areas on their own dedicated local networks as if they were all on the same network through the use of *ROS 2 Router*.

> This tutorial explains how to debug the *ROS 2 Router* to detect errors in the network configuration that may cause the discovery of the ROS 2 nodes and the deployed *ROS 2 Routers* to not be discovered and start communicating automatically.

Reviewer: ```suggestion
This tutorial explains how to debug the *ROS 2 Router* to detect errors in the network configuration that may cause the discovery of the ROS 2 nodes and the deployed *ROS 2 Routers* to not be discovered, preventing the correct behaviour of automatic data routing.
```

    Source the following file to setup the Vulcanexus environment:

    .. code-block:: bash

            source /opt/vulcanexus/humble/setup.bash

Running ROS 2 nodes
^^^^^^^^^^^^^^^^^^^

Once the environment has been setup using one of the above options, run the ROS 2 ``talker`` node in one terminal.

.. code-block:: bash

    ros2 run demo_nodes_cpp talker

Then, on another terminal, run the ROS 2 ``listener`` node in ROS 2 Domain ``1``.

.. code-block:: bash

    ROS_DOMAIN_ID=1 ros2 run demo_nodes_cpp listener

At this point, the ``listener`` should not receive any data.
If not, please go again through the previous steps.

Deploy ROS 2 Router
-------------------

Then, create the ROS 2 Router configuration file as the one shown below.

.. note::

    If deploying Vulcanexus from the Docker image, note that you will need to have a configuration file (config.yaml) for the DDS Router Edge accessible from your Docker container.

    This can be achieved by mounting a shared volume when launching the container, by copying the file from the local host to the container in case it is already running, or by editing a file from the Docker container itself.

> .. literalinclude:: /resources/tutorials/cloud/change_domain/change_domain_with_echo.yaml

Reviewer: NIT: add a button to download the yaml file?

Author: I'm not sure if we can do that

Reviewer 2: Kind of in the same line of thought: https://github.com/eProsima/vulcanexus/pull/86#discussion_r1018947827

Reviewer: It is possible to do it explicitly: https://fast-dds.docs.eprosima.com/en/latest/fastdds/ros2/discovery_server/ros2_discovery_server.html#compare-discovery-server-with-simple-discovery
However I am not sure we would want that

    :language: yaml

Main aspect of this configuration is the Echo Participant.
This type of participant, specific for debugging purposes, is explained in detail below.

Echo Participant
^^^^^^^^^^^^^^^^

> The `Echo Participant` is a participant/interface of the *ROS 2 Router* that prints in ``stdout`` all discovery information and/or user data that is received by the *ROS 2 Router*. Therefore, this participant does not perform any discovery o data reception functionality.

Reviewer: ```suggestion
The `Echo Participant` is a participant/interface of the *ROS 2 Router* that prints in ``stdout`` all discovery information and/or user data that is received by the *ROS 2 Router*. Therefore, this participant does not perform any discovery or data reception functionality.
```

In the case of discovery traces, messages such as the following will be displayed:

.. code-block:: bash

>     New endpoint discovered: Endpoint{<endpoint_guid>;<endpoint_kind>;<topic>}.

Reviewer: ```suggestion
    New endpoint discovered: Endpoint{<endpoint_guid>;<endpoint_kind>;<topic_info>}.
```

For data reception messages, the traces show the following information:

.. code-block:: bash

    Received data in Participant: <participant_id> in topic: <topic>.

These logs contain the `Participant Name` of the participant that has originally received the message, and the

> `Topic` where this message has been received.

Reviewer: ```suggestion
`Topic Information` where this message has been received. The `Topic Information` shown the `Topic Name`, `Topic Data Type Name` and the `QoS`.
```

> Additionally, extra information such as the data `Payload` (in hexadecimal format) and source `Endpoint` `Guid` is displayed in verbose mode:

Reviewer: ```suggestion
Additionally, extra information such as the data `Payload` (in hexadecimal format) and source `Endpoint` `Guid` is displayed in ``verbose`` mode:
```

    .. code-block:: bash

        mkdir -p ddsrouter_ws/src
        cd ddsrouter_ws
        git clone https://github.com/eProsima/DDS-Router src/ddsrouter
        colcon build --cmake-args -DLOG_INFO=ON

Running ROS 2 Router
^^^^^^^^^^^^^^^^^^^^

Update the environment setup to use the built ROS 2 Router instead of the one delivered in Vulcanexus.

.. code-block:: bash

    source ddsrouter_ws/install/setup.bash

Monitor ROS 2 Router internal operation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

ROS 2 Router offers several input arguments to configure the information displayed when it is built with the internal debug traces.
These are:

-   ``--log-verbosity <info|warning|error>``: set the verbosity level so only log messages with equal or higher importance level are shown.
-   ``--log-filer <regex>``: set a regex string as filter.
    Only log messages with a category that matches this regex will be printed (``ERROR`` messages will be always shown unless ``log-verbosity`` argument is set to ``ERROR``).
-   ``--debug``: set ``log-verbosity`` to ``info`` and ``log-filter`` to ``DDSROUTER``.

Thus, ROS 2 Router can be run as follows:

.. code-block::

>     ddsrouter -c <path/to/file>/echo.yaml --log-verbosity warning --log-filer DDSROUTER_DISCOVERY_DATABASE

Reviewer: Please, add an example of a trace of a router execution.
e.g. 
`2022-11-14 09:42:36.735 [DDSROUTER_DISCOVERY_DATABASE Info] Creating queue processing thread. -> Function DiscoveryDatabase`
